core/drivers/crypto/hisilicon/sec_main.c

```diff
@@ -0,0 +1,288 @@
+{
+	struct acc_device *sec_dev = NULL;
+	struct acc_device *cur_dev = NULL;
+	uint32_t free_qp_num = 0;
+	uint32_t max_qp_num = 0;
+
+	SLIST_FOREACH(cur_dev, &sec_list, link) {
+		if (cur_dev->qm.fun_type == HISI_QM_HW_PF)
+			free_qp_num = HISI_QM_PF_Q_NUM - cur_dev->qm.qp_in_used;
+		else
+			free_qp_num = HISI_QM_VF_Q_NUM - cur_dev->qm.qp_in_used;
+		if (free_qp_num > max_qp_num) {
+			max_qp_num = free_qp_num;
+			sec_dev = cur_dev;
+		}
+	}
+
+	if (!sec_dev) {
+		EMSG("No available sec device!\n");
+		return NULL;
+	}
+
+	return hisi_qm_create_qp(&sec_dev->qm, sq_type);
+}
+
+static void sec_disable_clock_gate(struct hisi_qm *qm)
+{
+	uint32_t val;
+
+	/* QM_HW_V2 version need to close clock gating */
+	val = io_read32(qm->io_base + SEC_CONTROL_REG);
+	val &= SEC_CLK_GATE_DISABLE;
+	io_write32(qm->io_base + SEC_CONTROL_REG, val);
+}
+
+static void sec_enable_clock_gate(struct hisi_qm *qm)
+{
+	uint32_t val;
+
+	if (qm->version == HISI_QM_HW_V2)
+		return;
+
+	val = io_read32(qm->io_base + SEC_CONTROL_REG);
+	val |= SEC_CLK_GATE_ENABLE;
+	io_write32(qm->io_base + SEC_CONTROL_REG, val);
+	io_write32(qm->io_base + SEC_DYNAMIC_GATE_V3, SEC_DYNAMIC_GATE_EN);
+	io_write32(qm->io_base + SEC_CORE_AUTO_GATE_V3, SEC_CORE_AUTO_GATE_EN);
+}
+
+static TEE_Result sec_engine_init(struct acc_device *sec_dev)
+{
+	struct hisi_qm *qm = &sec_dev->qm;
+	uint32_t val = 0;
+	int32_t ret;
+
+	if (qm->fun_type == HISI_QM_HW_VF)
+		return HISI_QM_DRVCRYPT_NO_ERR;
+
+	sec_disable_clock_gate(qm);
+	hisi_qm_dev_init(qm);
+
+	io_write32(qm->io_base + SEC_MEM_START_INIT, 0x1);
+	ret = IO_READ32_POLL_TIMEOUT(qm->io_base + SEC_MEM_INIT_DONE, val,
+				     val & 0x1, POLL_PERIOD, POLL_TIMEOUT);
+	if (ret) {
+		EMSG("Fail to init sec mem!\n");
+		return ret;
+	}
+
+	val = io_read32(qm->io_base + SEC_CONTROL_REG);
+	io_setbits32(qm->io_base + SEC_CONTROL_REG, sec_dev->endian);
+
+	if (qm->version == HISI_QM_HW_V2) {
+		/* smmu bypass */
+		io_write32(qm->io_base + SEC_INTERFACE_USER_CTRL0,
+			   SEC_USER0_CFG);
+		io_write32(qm->io_base + SEC_INTERFACE_USER_CTRL1,
+			   SEC_USER1_CFG);
+		io_write32(qm->io_base + AM_CFG_SINGLE_PORT_MAX_TRANS,
+			   SEC_SINGLE_PORT_MAX_TRANS);
+		io_write32(qm->io_base + SEC_SAA_EN, SEC_SAA_ENABLE);
+		/* HW V2 enable sm4 extra mode, as ctr/ecb */
+		io_write32(qm->io_base + SEC_BD_ERR_CHK_EN0, SEC_BD_ERR_CHK0);
+		/* Enable sm4 xts mode multiple iv */
+		io_write32(qm->io_base + SEC_BD_ERR_CHK_EN1, SEC_BD_ERR_CHK1);
+		/* disable pbkdf2 len check */
+		io_write32(qm->io_base + SEC_BD_ERR_CHK_EN2, SEC_BD_ERR_CHK2);
+		io_write32(qm->io_base + SEC_BD_ERR_CHK_EN3, SEC_BD_ERR_CHK3);
+	} else {
+		/* cmd_type is controlled by hac subctrl, default normal */
+		io_write32(qm->io_base + SEC_INTERFACE_USER_CTRL0_V3,
+			   SEC_USER0_CFG_V3);
+		io_write32(qm->io_base + SEC_INTERFACE_USER_CTRL1_V3,
+			   SEC_USER1_CFG_V3);
+		io_write32(qm->io_base + SEC_SAA_EN, SEC_SAA_ENABLE_V3);
+		/* disable pbkdf2 salt len check */
```

Comment:
`/* Disable PBKDF2 salt length check */`

Reply:
Thank you for your comments, I have made changes based on your comments.